### PR TITLE
Added verify transactions management command to ensure orders and pay…

### PR DESCRIPTION
**Pull request by calistenson**
…ments match.

https://openedx.atlassian.net/browse/LEARNER-5763
https://openedx.atlassian.net/browse/LEARNER-6012

### CODE REVIEW

File: ecommerce/core/management/commands/verify_transactions.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+Order = get_model('order', 'Order')
+PaymentEvent = get_model('order', 'PaymentEvent')
+
+
+class Command(BaseCommand):
+
+    help = 'Management command to verify ecommerce transactions and trigger alerts if there is any imbalance.'
+
+    def handle(self, *args, **options):
+        exit_errors = {}
```
Comment (BenjiLee) [inline ecommerce/core/management/commands/verify_transactions.py]: Add a docstring.

- An example or structure of the output.
- What exceptions are handled

All of this might be better off in the main docstring for the file, but it should be noted here or at the top of the file.

File: ecommerce/core/management/commands/verify_transactions.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
+"""
+Django management command to verify ecommerce transactions.
```
Comment (BenjiLee) [inline ecommerce/core/management/commands/verify_transactions.py:2]: Add more details about the management command as a whole.

File: ecommerce/core/management/commands/verify_transactions.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+from oscar.core.loading import get_model
+
+logger = logging.getLogger(__name__)
+Order = get_model('order', 'Order')
+PaymentEvent = get_model('order', 'PaymentEvent')
+
+
+class Command(BaseCommand):
+
+    help = 'Management command to verify ecommerce transactions and trigger alerts if there is any imbalance.'
```
Comment (BenjiLee) [inline ecommerce/core/management/commands/verify_transactions.py]: Since we are not explicitly triggering alerts, we should clarify that we are just outputting logs.

File: ecommerce/core/management/commands/verify_transactions.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        order_sum = 0
+        payment_sum = 0
+
+        for order in orders:
+            order_sum += order.total_incl_tax
+            pays = PaymentEvent.objects.filter(order_id=order.id)
+            if pays.count() == 0 and order.total_incl_tax > 0:
+                orders_without_payments.append(order.id)
+            elif pays.count() > 1:
+                exit_errors["multi_payment"] = "Multiple payments on order."
```
Comment (BenjiLee) [inline ecommerce/core/management/commands/verify_transactions.py]: Add more details to this log message such as which orders and what payments.

File: ecommerce/core/management/commands/verify_transactions.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        payments = PaymentEvent.objects.all().filter(date_created__gte=start, date_created__lt=end)
+
+        orders_without_payments = []
+
+        # Sum the total order costs and payment costs then verify they're equal.
+        order_sum = 0
+        payment_sum = 0
+
+        for order in orders:
+            order_sum += order.total_incl_tax
```
Comment (BenjiLee) [inline ecommerce/core/management/commands/verify_transactions.py]: I think every order after the first erroneous order will fail because we are not resetting the sums.

File: ecommerce/core/management/commands/verify_transactions.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+            pays = PaymentEvent.objects.filter(order_id=order.id)
```
Comment (BenjiLee) [inline ecommerce/core/management/commands/verify_transactions.py]: We should use a better variable name here such as `payment_events` or maybe `payment_events_for_order`

Comment (BenjiLee) [reply ecommerce/core/management/commands/verify_transactions.py]: This filters against the full database. I think you were trying to further filter against your queryset that you filtered in the previous lines. (`payments`)

File: ecommerce/core/management/commands/verify_transactions.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+            if pays.count() == 0 and order.total_incl_tax > 0:
```
Comment (BenjiLee) [inline ecommerce/core/management/commands/verify_transactions.py]: if an order.total_incl_tax == 0, do we want to verify that there is no payment event?

File: ecommerce/core/management/commands/verify_transactions.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+                orders_without_payments.append(order.id)
+            elif pays.count() > 1:
+                exit_errors["multi_payment"] = "Multiple payments on order."
+
+        for payment in payments:
+            payment_sum += payment.amount
+
+        if order_sum != payment_sum:
+            exit_errors["totals_mismatch"] = "Order totals mismatch with payments received." \
```
Comment (BenjiLee) [inline ecommerce/core/management/commands/verify_transactions.py]: When there is a mismatch, how can we identify which orders/paymentevents are wrong if  all orders have payments.

**Review by BenjiLee — CHANGES_REQUESTED** [review]

File: ecommerce/core/management/commands/tests/test_verify_transactions.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+PaymentEventTypeName = get_class('order.constants', 'PaymentEventTypeName')
+
+
+class VerifyTransactionsTest(TestCase):
+
+    def setUp(self):
+        self.payevent = PaymentEventType.objects.get_or_create(name=PaymentEventTypeName.PAID)
+        self.timestamp = datetime.datetime.now(pytz.utc) - datetime.timedelta(hours=1, minutes=30)
+
+    def tearDown(self):
```
Comment (BenjiLee) [inline ecommerce/core/management/commands/tests/test_verify_transactions.py]: We should be able to get rid of this. All changes are rolled back after the test.

https://docs.djangoproject.com/en/1.9/topics/testing/tools/#transactiontestcase

File: ecommerce/core/management/commands/tests/test_verify_transactions.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+        Order.objects.all().delete()
+
+    def test_no_errors(self):
+        """Test verify_transactions with order and payment of same amount."""
+        order = OrderFactory()
```
Comment (BenjiLee) [inline ecommerce/core/management/commands/tests/test_verify_transactions.py]: I believe we can shorten this to...

`OrderFactory(total_incl_tax = 100, date_placed = self.timestamp)`
```
order = OrderFactory()
order.total_incl_tax = 100
order.date_placed = self.timestamp
```

File: ecommerce/core/management/commands/tests/test_verify_transactions.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+        order.total_incl_tax = 100
+        order.date_placed = self.timestamp
+        order.save()
+        payment = PaymentEventFactory(order=order, amount=100, event_type_id=self.payevent[0].id)
```
Comment (BenjiLee) [inline ecommerce/core/management/commands/tests/test_verify_transactions.py]: Same refactoring comment as Order() above.

File: ecommerce/core/management/commands/tests/test_verify_transactions.py [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+        payment.date_created = self.timestamp
+        payment.save()
+        try:
+            call_command('verify_transactions')
+        except CommandError:
+            self.fail("Failed to verify transactions when no errors were expected.")
+
+    def test_no_payment_for_order(self):
+        """Verify errors are thrown when there are orders without payments."""
+        order = OrderFactory()
```
Comment (BenjiLee) [inline ecommerce/core/management/commands/tests/test_verify_transactions.py]: Looks like we are using a lot of the same code for the OrderFactory(). We can put this code in setUp().

**Comment by BenjiLee** [comment]
Git suggests using present-tense and imperative language. https://git.kernel.org/pub/scm/git/git.git/tree/Documentation/SubmittingPatches?id=HEAD#n133
`Added verify transactions management command to ensure orders and payments match.` -> `Add verify_transactions management command to ensure orders and payments match.`

Shorten the first line of the commit message. If it need more details, begin the paragraph on the 3rd line.
```
Add verify_transactions management command

The verify_transactions management command ensures that
orders amounts and paymentevent amounts match.

LEARNER-5763
LEARNER-6012
```

File: ecommerce/core/management/commands/verify_transactions.py [hunk 13]
```diff
@@ -0,0 +1,65 @@
+"""
+Django management command to verify ecommerce transactions.
+
+Command is run by Jenkins job each hour on the previous hour and 10 minutes of orders in ecommerce.
+
+For each order in the time window the command verifies exactly one payment of the expected value exists in the database.
```
Comment (BenjiLee) [inline ecommerce/core/management/commands/verify_transactions.py]: Nit and FYI
https://edx.readthedocs.io/projects/edx-developer-guide/en/latest/style_guides/python-guidelines.html#id4

Docstrings should be 80 chars in width.

File: ecommerce/core/management/commands/verify_transactions.py [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+
+If a PaymentEvent does not exist, multiple payments exist, or the payment amount is different from the order amount
```
Comment (BenjiLee) [inline ecommerce/core/management/commands/verify_transactions.py]:
```
If a PaymentEvent does not exist, multiple PaymentEvents exist, or the
PaymentEvent amount is different from the order amount, then the order 
id and relevant payment information is logged in a list associated with
each of these scenarios.
```

**Comment by BenjiLee** [comment]
@calistenson Could you output an example of this management command where all 3 scenarios fail at the same time?

File: ecommerce/core/management/commands/tests/test_verify_transactions.py [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+Order = get_model('order', 'Order')
+PaymentEvent = get_model('order', 'PaymentEvent')
+PaymentEventType = get_model('order', 'PaymentEventType')
+PaymentEventTypeName = get_class('order.constants', 'PaymentEventTypeName')
+
+
+class VerifyTransactionsTest(TestCase):
+
+    def setUp(self):
+        self.payevent = PaymentEventType.objects.get_or_create(name=PaymentEventTypeName.PAID)
```
Comment (BenjiLee) [inline ecommerce/core/management/commands/tests/test_verify_transactions.py]: nit: a `payevent` is not a `paymenteventtype`.`self.payevent` -> `self.payment_event_type`

File: ecommerce/core/management/commands/tests/test_verify_transactions.py [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+        self.timestamp = datetime.datetime.now(pytz.utc) - datetime.timedelta(hours=1, minutes=30)
```
Comment (BenjiLee) [inline ecommerce/core/management/commands/tests/test_verify_transactions.py]: nit: self.timestamp is not a random timestamp. This might be easier to code when we have parser arguments being passed in for the management command. As it is right now, changing the management command time window could potentially break tests. Having a way to pass in a time window will make our tests more flexible.

File: ecommerce/core/management/commands/tests/test_verify_transactions.py [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+        self.order = OrderFactory(total_incl_tax=90, date_placed=self.timestamp)
+        self.order.save()
+
+    def test_no_errors(self):
+        """Test verify_transactions with order and payment of same amount."""
+        payment = PaymentEventFactory(order=self.order, amount=90, event_type_id=self.payevent[0].id)
+        payment.date_created = self.timestamp
```
Comment (BenjiLee) [inline ecommerce/core/management/commands/tests/test_verify_transactions.py]: Nit: Should be able to shove this into the factory initialization.
```
payment = PaymentEventFactory(
            order=self.order, 
            amount=90, 
            event_type_id=self.payevent[0].id, 
            date_created=self.timestamp
        )
```
this can also be done to the payment events below

**Comment by BenjiLee** [comment]
@calistenson  Is the same process going to be followed for a refunded order?

File: ecommerce/core/management/commands/verify_transactions.py [hunk 18]
```diff
@@ -0,0 +1,88 @@
+"""
+Django management command to verify ecommerce transactions.
+
+Command is run by Jenkins job each hour on the previous hour and 10 minutes of orders in ecommerce.
```
Comment (BenjiLee) [inline ecommerce/core/management/commands/verify_transactions.py]: This is no longer true. We can make this more resistant to changes.

`Command will be run as a cron job in Jenkins that looks at a time window for Order/Payment discrepancies.`

File: ecommerce/core/management/commands/verify_transactions.py [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+
+        orders = Order.objects.all().filter(date_placed__gte=start, date_placed__lt=end)
+
+        orders_without_payments = []
+        multi_payment_on_order = []
+        order_payment_totals_mismatch = []
+
+        for order in orders:
+            payment_events = PaymentEvent.objects.filter(order_id=order.id)
+            if payment_events.count() == 0 and order.total_incl_tax > 0:
```
Comment (BenjiLee) [inline ecommerce/core/management/commands/verify_transactions.py:82]: Add a comment to why this conditional is here. 
`# If a coupon is used to purchase a product for the full price, there will be no PaymentEvent`

Comment (calistenson) [reply ecommerce/core/management/commands/verify_transactions.py:82]: Done

File: ecommerce/core/management/commands/verify_transactions.py [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+            if payment_events.count() > 1:
+                if payment_events.count() == 2:
+                    event1 = payment_events[0]
+                    event2 = payment_events[1]
+                    if event1.event_type_id == event2.event_type_id:
+                        # If PaymentEvents are the same type then there were multiple payments.
+                        multi_payment_on_order.append((order.id, [event.id for event in payment_events]))
+                else:
+                    multi_payment_on_order.append((order.id, [event.id for event in payment_events]))
+            for event in payment_events:
```
Comment (BenjiLee) [inline ecommerce/core/management/commands/verify_transactions.py:94]: nit: I believe we should be guaranteed only 1 payment event here. We are using this for loop to extract a single item from a list and rename it.

File: ecommerce/core/management/commands/verify_transactions.py [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+                print("event: ", event)
+                if event.amount != order.total_incl_tax:
+                    order_payment_totals_mismatch.append(("Order: " + str(order.id) + " Amount: " +
+                                                          str(order.total_incl_tax), "Payment: " +
+                                                          str(event.id) + " Amount: " +
+                                                          str(event.amount)))
+
+        if len(order_payment_totals_mismatch):
```
Comment (BenjiLee) [inline ecommerce/core/management/commands/verify_transactions.py]: fun python! empty lists are falsy

File: ecommerce/core/management/commands/tests/test_verify_transactions.py [hunk 22]
```diff
@@ -0,0 +1,137 @@
...
+                                       event_type_id=self.payevent[0].id,
+                                       date_created=self.timestamp)
+        payment1.save()
+        payment2.save()
+        with self.assertRaises(CommandError) as cm:
+            call_command('verify_transactions')
+        exception = cm.exception
+        self.assertIn("The following orders had multiple payments ", exception.message)
+        self.assertIn(str(self.order.id), exception.message)
+        self.assertIn(str(payment1.id), exception.message)
```
Comment (BenjiLee) [inline ecommerce/core/management/commands/tests/test_verify_transactions.py:94]: There's a potential for order.id == payment.id since they are not unique with each other. A way to resolve this is to customize the id generation field in the PaymentFactory. Example: https://github.com/edx/ecommerce/blob/af0190314929a482421f52162ccad3eaf296e982/ecommerce/courses/tests/factories.py#L14

Since we are running tests in a fresh test database, the likelyhood of order.id == payment.id is high.

File: ecommerce/core/management/commands/verify_transactions.py (hunk 20 above)
Comment (calistenson) [reply ecommerce/core/management/commands/verify_transactions.py:94]: We still want to catch mismatched totals in the cases where there are multiple payments. So I will leave the loop here.

File: ecommerce/core/management/commands/tests/test_verify_transactions.py (hunk 22 above)
Comment (calistenson) [reply ecommerce/core/management/commands/tests/test_verify_transactions.py:94]: Fixed this issue in the test by setting the order id manually to "1000". Should I add a comment in the test about why I did this?

File: ecommerce/core/management/commands/tests/factories.py [hunk 23]
```diff
@@ -0,0 +1,8 @@
+import factory
+from oscar.core.loading import get_model
+
+
+class PaymentEventFactory(factory.DjangoModelFactory):
+
```
Comment (BenjiLee) [inline ecommerce/core/management/commands/tests/factories.py:7]: `id = random.randint(1000,9999)`

This it basically an int version of FuzzyText for our purposes

**Review by BenjiLee — APPROVED** [review]

**Comment by BenjiLee** [comment]
shipit
